src/mongocxx/test/v1/read_preference.cpp

```diff
@@ -0,0 +1,238 @@
+
+//
+
+#include <bsoncxx/test/v1/document/view.hh>
+#include <bsoncxx/test/v1/stdx/string_view.hh>
+
+#include <chrono>
+#include <tuple>
+#include <utility>
+
+#include <bsoncxx/private/suppress_deprecation_warnings.hh>
+
+#include <mongocxx/private/scoped_bson.hh>
+
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/generators/catch_generators.hpp>
+
+namespace mongocxx {
+namespace v1 {
+
+TEST_CASE("ownership", "[mongocxx][v1][read_preference]") {
+    using mode = read_preference::read_mode;
+
+    read_preference source;
+    read_preference target;
+
+    source.mode(mode::k_primary_preferred);
+    target.mode(mode::k_secondary_preferred);
+
+    SECTION("move") {
+        auto move = std::move(source);
+
+        // source is in an assign-or-move-only state.
+
+        CHECK(move.mode() == mode::k_primary_preferred);
+
+        target = std::move(move);
+
+        // source is in an assign-or-move-only state.
+
+        CHECK(target.mode() == mode::k_primary_preferred);
+    }
+
+    SECTION("copy") {
+        auto copy = source;
+
+        CHECK(source.mode() == mode::k_primary_preferred);
+        CHECK(copy.mode() == mode::k_primary_preferred);
+
+        target = copy;
+
+        CHECK(copy.mode() == mode::k_primary_preferred);
+        CHECK(target.mode() == mode::k_primary_preferred);
+    }
+}
+
+TEST_CASE("default", "[mongocxx][v1][read_preference]") {
+    using mode = read_preference::read_mode;
+
+    read_preference const rp;
+
+    BSONCXX_SUPPRESS_DEPRECATION_WARNINGS_BEGIN
+    CHECK_FALSE(rp.hedge().has_value());
+    BSONCXX_SUPPRESS_DEPRECATION_WARNINGS_END
+
+    CHECK(rp.mode() == mode::k_primary);
+    CHECK(rp.tags().empty());
+    CHECK_FALSE(rp.max_staleness().has_value());
+}
+
+TEST_CASE("hedge", "[mongocxx][v1][read_preference]") {
+    BSONCXX_SUPPRESS_DEPRECATION_WARNINGS_BEGIN
+
+    read_preference rp;
+
+    auto const v = GENERATE(values({
+        scoped_bson{},
+        scoped_bson{R"({})"},
+        scoped_bson{R"({"enabled": false})"},
+        scoped_bson{R"({"enabled": true})"},
+    }));
+
+    CHECK_NOTHROW(rp.hedge(v.view()));
+
+    CHECKED_IF(v.view().empty()) {
+        CHECK_FALSE(rp.hedge().has_value());
+    }
+
+    else {
+        CHECK(rp.hedge() == v.view());
+    }
+
+    BSONCXX_SUPPRESS_DEPRECATION_WARNINGS_END
+}
+
+TEST_CASE("mode", "[mongocxx][v1][read_preference]") {
+    using mode = read_preference::read_mode;
+
+    mode input = {};
+    mode expected = {};
+
+    std::tie(input, expected) = GENERATE(
+        table<mode, mode>({
+            {mode::k_primary, mode::k_primary},
+            {mode::k_primary_preferred, mode::k_primary_preferred},
+            {mode::k_secondary, mode::k_secondary},
+            {mode::k_secondary_preferred, mode::k_secondary_preferred},
+            {mode::k_nearest, mode::k_nearest},
+            {static_cast<mode>(-1), mode::k_primary},
+            {static_cast<mode>(5), mode::k_primary},
```

Comment:
Prior to this PR, using a value outside of the enum hits a [BSON_UNREACHABLE](https://github.com/mongodb/mongo-cxx-driver/blob/29185e4c3f21ce6d192ad55870207e96df765e7f/src/mongocxx/lib/mongocxx/v_noabi/mongocxx/conversions.cpp#L37):

```c++
// Prior to this PR:
TEST_CASE("Bad read mode", "[read_preference]") {
    read_preference rp;
    rp.mode(static_cast<mongocxx::v_noabi::read_preference::read_mode>(-1)); // Abort!
}
```

Suggest restoring this behavior (or throwing an exception?). That may help add a future `read_mode` enumerator without concern for breaking consumers. E.g. if consumers are relying on `5` meaning primary, adding a new read mode might later break that assumption.